File: ui/src/lib/installSource.ts
export type InstallSource =
  | { kind: "git-url"; value: string }
  | { kind: "github"; owner: string; repo: string; ref?: string }
  | { kind: "plugin"; id: string }
  | { kind: "folder"; value: string }
  | { kind: "npx-skills"; source: string; skill?: string };

const GITHUB_SHORTHAND = /^([A-Za-z0-9_.-]+)\/([A-Za-z0-9_.-]+)(?:@([A-Za-z0-9._/-]+))?$/;
const PLUGIN_ID = /^[^/\s\\]+@[^/\s\\]+$/;
const HTTPS_GIT = /^https:\/\/[^\s]+\/[^\s]+$/i;
const INVALID = "Use an HTTPS git URL, owner/repo, name@marketplace, or npx skills add.";

export function parseInstallSource(input: string): InstallSource | { error: string } {
  const value = input.trim();
  if (!value) {
    return { error: INVALID };
  }
  if (/^(git@|ssh:\/\/)/i.test(value)) {
    return { error: INVALID };
  }
  const npx = parseNpxSkills(value);
  if (npx) {
    return npx;
  }
  if (HTTPS_GIT.test(value)) {
    return { kind: "git-url", value };
  }
  if (looksLikeAbsPath(value)) {
    return { kind: "folder", value };
  }
  if (PLUGIN_ID.test(value)) {
    return { kind: "plugin", id: value };
  }
  const shorthand = GITHUB_SHORTHAND.exec(value);
  if (shorthand) {
    return {
      kind: "github",
      owner: shorthand[1],
      repo: shorthand[2],
      ref: shorthand[3],
    };
  }
  return { error: INVALID };
}

export function isValidInstallInput(text: string): boolean {
  return !("error" in parseInstallSource(text));
}

export function resolvedInstallSource(text: string): string | null {
  const parsed = parseInstallSource(text);
  if ("error" in parsed) {
    return null;
  }
  if (parsed.kind === "github") {
    return parsed.ref ? `${parsed.owner}/${parsed.repo}@${parsed.ref}` : `${parsed.owner}/${parsed.repo}`;
  }
  if (parsed.kind === "plugin") {
    return parsed.id;
  }
  if (parsed.kind === "npx-skills") {
    return parsed.skill
      ? `npx skills add ${parsed.source} --skill ${parsed.skill}`
      : `npx skills add ${parsed.source}`;
  }
  return parsed.value;
}

export function installHint(parsed: InstallSource | { error: string }): string {
  if ("error" in parsed) {
    return "HTTPS git, owner/repo, name@marketplace, or npx skills add. SSH later.";
  }
  switch (parsed.kind) {
    case "plugin":
      return "Install this plugin from a registered marketplace.";
    case "github":
      return "Add this GitHub repo as a marketplace. Then install plugins with name@marketplace.";
    case "git-url":
      return "Add this git repo as a marketplace.";
    case "folder":
      return "Link a local marketplace or plugin folder.";
    case "npx-skills":
      return parsed.skill
        ? `Install skill “${parsed.skill}” globally via npx.`
        : "Install skills globally via npx skills add.";
  }
}

function parseNpxSkills(value: string): InstallSource | { error: string } | null {
  const tokens = tokenize(value);
  const first = tokens[0];
  if (!first) {
    return null;
  }
  const npx = first.toLowerCase() === "npx";
  if (!npx && first.toLowerCase() !== "skills") {
    return null;
  }
  let index = npx ? 1 : 0;
  if (npx) {
    while (index < tokens.length && isNpxPrefixFlag(tokens[index])) {
      index += 1;
    }
  }
  if (tokens[index]?.toLowerCase() !== "skills") {
    return { error: INVALID };
  }
  index += 1;
  if (tokens[index]?.toLowerCase() !== "add") {
    return { error: INVALID };
  }
  index += 1;
  let source: string | undefined;
  let skill: string | undefined;
  while (index < tokens.length) {
    const token = tokens[index];
    if (token === "--skill" || token === "-s") {
      index += 1;
      skill = tokens[index];
    } else if (token === "-a" || token === "--agent" || token === "--agents") {
      index += 1;
    } else if (!isSkillsIgnoreFlag(token) && !token.startsWith("-") && !source) {
      source = token;
    }
    index += 1;
  }
  if (!source) {
    return { error: INVALID };
  }
  if (/^(git@|ssh:\/\/)/i.test(source)) {
    return { error: INVALID };
  }
  return skill ? { kind: "npx-skills", source, skill } : { kind: "npx-skills", source };
}

function tokenize(value: string): string[] {
  const tokens: string[] = [];
  let current = "";
  let quote: string | null = null;
  for (const ch of value) {
    if (quote) {
      if (ch === quote) {
        quote = null;
      } else {
        current += ch;
      }
      continue;
    }
    if (ch === '"' || ch === "'") {
      quote = ch;
      continue;
    }
    if (/\s/.test(ch)) {
      if (current) {
        tokens.push(current);
        current = "";
      }
      continue;
    }
    current += ch;
  }
  if (current) {
    tokens.push(current);
  }
  return tokens;
}

function isNpxPrefixFlag(token: string): boolean {
  return token === "-y" || token === "--yes" || token === "-q" || token === "--quiet";
}

function isSkillsIgnoreFlag(token: string): boolean {
  return (
    token === "-g" ||
    token === "--global" ||
    token === "-y" ||
    token === "--yes" ||
    token === "--all" ||
    token === "--copy" ||
    token === "-l" ||
    token === "--list"
  );
}

function looksLikeAbsPath(value: string): boolean {
  return /^[A-Za-z]:[\\/]/.test(value) || value.startsWith("\\\\") || value.startsWith("/");
}

File: src-tauri/src/install_source.rs
use std::path::{Path, PathBuf};

use crate::dto::AdapterError;

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum InstallSource {
    GitUrl(String),
    GitHub {
        owner: String,
        repo: String,
        ref_name: Option<String>,
    },
    LocalDir(PathBuf),
    Plugin(String),
    NpxSkills {
        source: String,
        skill: Option<String>,
    },
}

pub fn parse_install_source(input: &str) -> Result<InstallSource, AdapterError> {
    let value = input.trim();
    if value.is_empty() {
        return Err(AdapterError::message(INVALID));
    }
    if value.starts_with("git@") || value.to_ascii_lowercase().starts_with("ssh://") {
        return Err(AdapterError::message(INVALID));
    }
    if let Some(parsed) = parse_npx_skills(value) {
        return parsed;
    }
    if is_https_git(value) {
        return Ok(InstallSource::GitUrl(value.to_string()));
    }
    if looks_like_abs_path(value) {
        let path = PathBuf::from(value);
        if !path.is_dir() {
            return Err(AdapterError::message(format!("Folder not found: {value}")));
        }
        return Ok(InstallSource::LocalDir(path));
    }
    if let Some(id) = parse_plugin_id(value) {
        return Ok(InstallSource::Plugin(id));
    }
    if let Some((owner, repo, ref_name)) = parse_github(value) {
        return Ok(InstallSource::GitHub {
            owner,
            repo,
            ref_name,
        });
    }
    Err(AdapterError::message(INVALID))
}

impl InstallSource {
    pub fn as_cli_source(&self) -> String {
        match self {
            Self::GitUrl(url) => url.clone(),
            Self::GitHub {
                owner,
                repo,
                ref_name: Some(ref_name),
            } => format!("{owner}/{repo}@{ref_name}"),
            Self::GitHub {
                owner,
                repo,
                ref_name: None,
            } => format!("{owner}/{repo}"),
            Self::LocalDir(path) => path.display().to_string(),
            Self::Plugin(id) => id.clone(),
            Self::NpxSkills { source, skill } => match skill {
                Some(skill) => format!("{source} --skill {skill}"),
                None => source.clone(),
            },
        }
    }

    pub fn is_npx_skills(&self) -> bool {
        matches!(self, Self::NpxSkills { .. })
    }

    pub fn npx_skills_argv(&self, agent: &str) -> Vec<String> {
        let Self::NpxSkills { source, skill } = self else {
            return Vec::new();
        };
        let mut args = vec![
            "-y".into(),
            "skills".into(),
            "add".into(),
            source.clone(),
            "-g".into(),
            "-y".into(),
            "-a".into(),
            agent.to_string(),
        ];
        if let Some(skill) = skill {
            args.push("--skill".into());
            args.push(skill.clone());
        }
        args
    }

    pub fn claude_install_argv(&self) -> Vec<String> {
        match self {
            Self::Plugin(id) => vec![
                "plugin".into(),
                "install".into(),
                "-s".into(),
                "user".into(),
                "-y".into(),
                id.clone(),
            ],
            Self::NpxSkills { .. } => self.npx_skills_argv("claude-code"),
            _ => vec![
                "plugin".into(),
                "marketplace".into(),
                "add".into(),
                "--scope".into(),
                "user".into(),
                self.as_cli_source(),
            ],
        }
    }

    pub fn claude_uninstall_argv(plugin_id: &str) -> Vec<String> {
        vec![
            "plugin".into(),
            "uninstall".into(),
            "-s".into(),
            "user".into(),
            "-y".into(),
            plugin_id.to_string(),
        ]
    }

    pub fn codex_install_argv(&self) -> Vec<String> {
        match self {
            Self::Plugin(id) => vec!["plugin".into(), "add".into(), "--json".into(), id.clone()],
            Self::NpxSkills { .. } => self.npx_skills_argv("codex"),
            Self::GitHub {
                owner,
                repo,
                ref_name: Some(ref_name),
            } => vec![
                "plugin".into(),
                "marketplace".into(),
                "add".into(),
                "--json".into(),
                format!("{owner}/{repo}"),
                "--ref".into(),
                ref_name.clone(),
            ],
            _ => vec![
                "plugin".into(),
                "marketplace".into(),
                "add".into(),
                "--json".into(),
                self.as_cli_source(),
            ],
        }
    }

    pub fn codex_uninstall_argv(plugin_id: &str) -> Vec<String> {
        vec![
            "plugin".into(),
            "remove".into(),
            "--json".into(),
            plugin_id.to_string(),
        ]
    }

    pub fn agy_install_argv(&self) -> Vec<String> {
        match self {
            Self::NpxSkills { .. } => self.npx_skills_argv("antigravity"),
            _ => vec!["plugin".into(), "install".into(), self.as_cli_source()],
        }
    }

    pub fn agy_uninstall_argv(plugin_id: &str) -> Vec<String> {
        let (name, _) = crate::paths::plugin_id_parts(plugin_id);
        vec!["plugin".into(), "uninstall".into(), name]
    }

    pub fn plugin_marketplace(plugin_id: &str) -> Option<&str> {
        plugin_id
            .split_once('@')
            .map(|(_, source)| source)
            .filter(|source| !source.is_empty() && !source.eq_ignore_ascii_case("local"))
    }

    pub fn claude_marketplace_update_argv(marketplace: &str) -> Vec<String> {
        vec![
            "plugin".into(),
            "marketplace".into(),
            "update".into(),
            marketplace.to_string(),
        ]
    }

    pub fn claude_update_argv(plugin_id: &str) -> Vec<String> {
        vec![
            "plugin".into(),
            "update".into(),
            "-s".into(),
            "user".into(),
            "-y".into(),
            plugin_id.to_string(),
        ]
    }

    pub fn codex_marketplace_upgrade_argv(marketplace: &str) -> Vec<String> {
        vec![
            "plugin".into(),
            "marketplace".into(),
            "upgrade".into(),
            "--json".into(),
            marketplace.to_string(),
        ]
    }

    pub fn codex_update_argv(plugin_id: &str) -> Vec<String> {
        vec![
            "plugin".into(),
            "add".into(),
            "--json".into(),
            plugin_id.to_string(),
        ]
    }
}

const INVALID: &str = "Use an HTTPS git URL, owner/repo, name@marketplace, or npx skills add.";

fn is_https_git(value: &str) -> bool {
    let lower = value.to_ascii_lowercase();
    lower.starts_with("https://") && value.contains('/') && !value.chars().any(char::is_whitespace)
}

fn looks_like_abs_path(value: &str) -> bool {
    Path::new(value).is_absolute()
        || (value.len() >= 3
            && value.as_bytes()[0].is_ascii_alphabetic()
            && value.as_bytes()[1] == b':'
            && (value.as_bytes()[2] == b'\\' || value.as_bytes()[2] == b'/'))
}

fn parse_npx_skills(value: &str) -> Option<Result<InstallSource, AdapterError>> {
    let tokens = tokenize(value);
    let first = tokens.first()?;
    let npx = first.eq_ignore_ascii_case("npx");
    if !npx && !first.eq_ignore_ascii_case("skills") {
        return None;
    }
    let mut index = if npx { 1 } else { 0 };
    if npx {
        while index < tokens.len() && is_npx_prefix_flag(&tokens[index]) {
            index += 1;
        }
    }
    if index >= tokens.len() || !tokens[index].eq_ignore_ascii_case("skills") {
        return Some(Err(AdapterError::message(INVALID)));
    }
    index += 1;
    if index >= tokens.len() || !tokens[index].eq_ignore_ascii_case("add") {
        return Some(Err(AdapterError::message(INVALID)));
    }
    index += 1;
    let mut source = None;
    let mut skill = None;
    while index < tokens.len() {
        let token = &tokens[index];
        if token == "--skill" || token == "-s" {
            index += 1;
            skill = tokens.get(index).cloned();
        } else if token == "-a" || token == "--agent" || token == "--agents" {
            index += 1;
        } else if !is_skills_ignore_flag(token) && !token.starts_with('-') && source.is_none() {
            source = Some(token.clone());
        }
        index += 1;
    }
    let Some(source) = source else {
        return Some(Err(AdapterError::message(INVALID)));
    };
    if source.starts_with("git@") || source.to_ascii_lowercase().starts_with("ssh://") {
        return Some(Err(AdapterError::message(INVALID)));
    }
    Some(Ok(InstallSource::NpxSkills { source, skill }))
}

fn tokenize(value: &str) -> Vec<String> {
    let mut tokens = Vec::new();
    let mut current = String::new();
    let mut quote = None;
    for ch in value.chars() {
        if let Some(mark) = quote {
            if ch == mark {
                quote = None;
            } else {
                current.push(ch);
            }
            continue;
        }
        if ch == '"' || ch == '\'' {
            quote = Some(ch);
            continue;
        }
        if ch.is_whitespace() {
            if !current.is_empty() {
                tokens.push(std::mem::take(&mut current));
            }
            continue;
        }
        current.push(ch);
    }
    if !current.is_empty() {
        tokens.push(current);
    }
    tokens
}

fn is_npx_prefix_flag(token: &str) -> bool {
    matches!(token, "-y" | "--yes" | "-q" | "--quiet")
}

fn is_skills_ignore_flag(token: &str) -> bool {
    matches!(
        token,
        "-g" | "--global" | "-y" | "--yes" | "--all" | "--copy" | "-l" | "--list"
    )
}

fn parse_plugin_id(value: &str) -> Option<String> {
    let (name, source) = value.split_once('@')?;
    if name.is_empty()
        || source.is_empty()
        || name.contains('/')
        || source.contains('/')
        || name.contains('\\')
        || source.contains('\\')
        || value.chars().any(char::is_whitespace)
    {
        return None;
    }
    Some(value.to_string())
}

fn parse_github(value: &str) -> Option<(String, String, Option<String>)> {
    let (base, ref_name) = match value.split_once('@') {
        Some((base, ref_name)) if !ref_name.is_empty() => (base, Some(ref_name.to_string())),
        Some(_) => return None,
        None => (value, None),
    };
    let (owner, repo) = base.split_once('/')?;
    if owner.is_empty()
        || repo.is_empty()
        || owner.contains('/')
        || repo.contains('/')
        || owner.contains('\\')
        || repo.contains('\\')
    {
        return None;
    }
    Some((owner.to_string(), repo.to_string(), ref_name))
}

#[cfg(test)]
mod tests;
